File: typescript/packages/core/src/resource/bin/bin.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { BinAccessor } from '../../accessor/bin.ts'
import { BIN_COMMANDS } from '../../commands/builtin/bin/index.ts'
import type { RegisteredCommand } from '../../commands/config.ts'
import { read as readCore } from '../../core/bin/read.ts'
import { readdir as readdirCore } from '../../core/bin/readdir.ts'
import { stat as statCore } from '../../core/bin/stat.ts'
import { stream as streamCore } from '../../core/bin/stream.ts'
import { BIN_OPS } from '../../ops/bin/index.ts'
import type { RegisteredOp } from '../../ops/registry.ts'
import type { FileStat } from '../../types.ts'
import { type PathSpec, ResourceName } from '../../types.ts'
import { BaseResource, type Resource } from '../base.ts'

export const BIN_PREFIX = '/bin'

/**
 * Read-only view resource backing the /bin mount. Renders the vfs
 * runtime's vocabulary (general commands + installed CLIs) as one stub
 * file per name on every read; holds no storage of its own, so
 * installs and uninstalls show immediately.
 */
export class BinViewResource extends BaseResource implements Resource {
  readonly kind = ResourceName.BIN
  readonly cachesReads = false
  // Stubs render from in-memory state, so stat() sizes by rendering:
  // cheap, no network, and never null.
  readonly sizesAlwaysKnown = true
  readonly accessor: BinAccessor

  constructor(entries: () => Map<string, Uint8Array>) {
    super()
    this.accessor = new BinAccessor(entries)
  }

  open(): Promise<void> {
    return Promise.resolve()
  }

  close(): Promise<void> {
    return Promise.resolve()
  }

  ops(): readonly RegisteredOp[] {
    return BIN_OPS
  }

  commands(): readonly RegisteredCommand[] {
    return BIN_COMMANDS
  }

  streamPath(path: PathSpec): AsyncIterable<Uint8Array> {
    return streamCore(this.accessor, path)
  }

  readFile(path: PathSpec): Promise<Uint8Array> {
    return readCore(this.accessor, path)
  }

  readdir(path: PathSpec): Promise<string[]> {
    return readdirCore(this.accessor, path)
  }

  stat(path: PathSpec): Promise<FileStat> {
    return statCore(this.accessor, path)
  }
}

File: python/mirage/resource/bin/bin.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Callable

from mirage.accessor.bin import BinAccessor
from mirage.commands.builtin.bin import COMMANDS
from mirage.ops.bin import OPS
from mirage.resource.base import BaseResource

BIN_PREFIX = "/bin"


class BinViewResource(BaseResource):
    """Read-only view resource backing the /bin mount.

    Renders the vfs runtime's vocabulary (general commands + installed
    CLIs) as one stub file per name on every read; holds no storage of
    its own, so installs and uninstalls show immediately.

    Args:
        entries (Callable[[], dict[str, bytes]]): renders the current
            stub table (name -> content).
    """
    accessor: BinAccessor

    name = "bin"
    # Stubs render from in-memory state, so stat() sizes by rendering:
    # cheap, no network, and never None.
    SIZES_ALWAYS_KNOWN: bool = True

    def __init__(self, entries: Callable[[], dict[str, bytes]]) -> None:
        super().__init__()
        self.accessor = BinAccessor(entries)
        for fn in COMMANDS:
            self.register(fn)
        for op in OPS:
            self.register_op(op)
